core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -271,18 +287,23 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     GroupOverview(groupId, protocolType.getOrElse(""))
   }
 
-  def initializeOffsets(offsets: collection.Map[TopicPartition, OffsetAndMetadata],
-                        pendingTxnOffsets: Map[Long, mutable.Map[TopicPartition, OffsetAndMetadata]]) {
+  def initializeOffsets(offsets: collection.Map[TopicPartition, CommitRecordMetadataAndOffset],
+                        pendingTxnOffsets: Map[Long, mutable.Map[TopicPartition, CommitRecordMetadataAndOffset]]) {
     this.offsets ++= offsets
     this.pendingTransactionalOffsetCommits ++= pendingTxnOffsets
   }
 
-  def completePendingOffsetWrite(topicPartition: TopicPartition, offset: OffsetAndMetadata) {
-    if (pendingOffsetCommits.contains(topicPartition))
-      offsets.put(topicPartition, offset)
+  def completePendingOffsetWrite(topicPartition: TopicPartition, offsetWithCommitRecordMetadata: CommitRecordMetadataAndOffset) {
+    if (pendingOffsetCommits.contains(topicPartition)) {
+      // TODO(apurva) : Is this check necessary? Is it ever possible that a transactional offset commit would
```

Comment:
I think this is possible but would be really rare, as it has to happen that:

1. normal commit message get append first, then txn commit message, but txn delayed produce gets added to the purgatory's watch list first then normal commit's delayed produce.
2. when fetch request advanced the hw, the `checkAndComplete` call first get to complete the txn commit's delayed produce, and get its data in `pendingTransactionalOffsetCommits`.
3. `handleTxnCompletion` kicks in immediately after 2), gets its offset in `offsets`.
4. the fetch request handler thread continues traversing the watch list, then tries to move the normal commit's delayed produce's corresponding offsets to `offsets`.